sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,295 @@
+//           later called on the bucket that *does* have the consistent version, we have to report any
+//           such item as "present" to give the upper layer a chance to look at it.
+//    * If the item does not exist elsewhere in the cache, report it as deleted.
+//
+// If an item has been moved by sharding, it's possible this will cause it to be reported
+// as deleted and re-created.
+//
+// Either B or C might be nonexistent; just treat those as empty.
+//
+// If we're the secondary, we shouldn't have to worry about concurrent access.
+func (s *StoragePackerV2) InvalidateItems(ctx context.Context, path string, newValue []byte) (present []*Item, deleted []*Item, err error) {
+	// Double-check that it's on our path, and figure out which bucket it should be
+	bucketKey, err := s.bucketKeyFromPath(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var replacementBucket *LockedBucket
+	bucketRemoved := false
+
+	if newValue == nil {
+		bucketRemoved = true
+		// Create a placeholder so we can run the rest of the logic
+		// against an empty collection
+		replacementBucket = s.newEmptyBucket(bucketKey)
+	} else {
+		// Fake up a storage entry for the bucket and decode it
+		storage := &logical.StorageEntry{
+			Value: newValue,
+		}
+		replacementBucket, err = s.DecodeBucket(storage)
+		if err != nil {
+			return nil, nil, errwrap.Wrapf("unable to decode replicated bucket: {{err}}", err)
+		}
+		if replacementBucket.Key != bucketKey {
+			return nil, nil, fmt.Errorf("decoded bucket key %q doesn't match path component %q", replacementBucket.Key, bucketKey)
+		}
+	}
+
+	present = make([]*Item, 0)
+
+	originalBucket, err := s.GetBucket(ctx, bucketKey)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem finding original bucket: {{err}}", err)
+	}
+	originalBucket.Lock()
+	defer originalBucket.Unlock()
+
+	// Any item included in the new bucket should be reported present.
+	if !replacementBucket.HasShards && replacementBucket.ItemMap != nil {
+		for k, v := range replacementBucket.ItemMap {
+			present = append(present, &Item{
+				ID:   k,
+				Data: v,
+			})
+		}
+	}
+
+	// Is the new bucket a replacement, or new?  If it's new, nothing to compare against.
+	if originalBucket.Key == replacementBucket.Key {
+		// Any item not present in the new bucket *might* be deleted, or
+		// it might be elsewhere (in which case we want to report that value instead,
+		// to make sure the correct version is being used.)
+		maybeDeleted := itemDifferenceBetween(originalBucket, replacementBucket)
+		var revisit []*Item
+		deleted, revisit, err = s.identifyItemsAbsentOrShadowed(bucketKey, maybeDeleted)
+		if err != nil {
+			// Return partial information?
+			return present, nil, err
+		}
+		present = append(present, revisit...)
+	}
+
+	// Swap the replacement bucket in to the cache (or delete)
+	cacheKey := s.GetCacheKey(bucketKey)
+	s.bucketsCacheLock.RLock()
```

Comment:
I think this just needs to be `s.bucketsCacheLock.Lock()` since the cache is actually edited while the lock is being held. Likewise below on 128, it'll need to be `s.bucketsCacheLock.Unlock()`.

Reply:
Good catch, thanks.